spring-integration-core/src/main/java/org/springframework/integration/json/JsonToObjectTransformer.java

```diff
@@ -83,21 +112,72 @@ public String getComponentType() {
 
 	@Override
 	protected Object doTransform(Message<?> message) {
+		MessageHeaders headers = message.getHeaders();
+		boolean removeHeaders = false;
+		ResolvableType valueType = obtainResolvableTypeFromHeadersIfAny(headers);
+
+		if (valueType != null) {
+			removeHeaders = true;
+		}
+		else {
+			valueType = this.targetType;
+		}
+
+		Object result;
 		try {
-			if (this.targetClass != null) {
-				return this.jsonObjectMapper.fromJson(message.getPayload(), this.targetClass);
-			}
-			else {
-				Object result = this.jsonObjectMapper.fromJson(message.getPayload(), message.getHeaders());
-				AbstractIntegrationMessageBuilder<Object> messageBuilder = this.getMessageBuilderFactory().withPayload(result)
-						.copyHeaders(message.getHeaders())
-						.removeHeaders(JsonHeaders.HEADERS.toArray(new String[3]));
-				return messageBuilder.build();
-			}
+			result = this.jsonObjectMapper.fromJson(message.getPayload(), valueType);
+
 		}
 		catch (IOException e) {
 			throw new UncheckedIOException(e);
 		}
+
+		if (removeHeaders) {
+			return getMessageBuilderFactory()
+					.withPayload(result)
+					.copyHeaders(headers)
+					.removeHeaders(JsonHeaders.HEADERS.toArray(new String[3]))
```

Comment:
I thought `[0]` is preferred now, not new, I know, but might as well since we're changing.

https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size

Reply:
Hm. Looks like my IDEA after some update has missed this warning 😢

Pushed the fix for `0`. More over it is right now `4` anyway 😄